R/stat-sina.r

```diff
@@ -0,0 +1,186 @@
+#' @param neighbour_limit if the samples within the same y-axis bin are more
+#' than neighbour_limit, the samples's X coordinates will be adjusted.
+#' @param method choose the method to spread the samples within the same
+#' neighbourhood along the x-axis. Available methods: "density",
+#' "neighbourhood" (can be abbreciated, e.g. "d"). See \code{Details}.
+#' @param adjust adjusts the bandwidth of the density kernel when
+#' \code{method == "density"} (see \code{\link[stats]{density}}) or the spread
+#' of the samples within the same group along the x-axis.
+#' @export
+stat_sina <-function(mapping = NULL, data = NULL,
+                     geom = "point", position = "identity",
+                     ...,
+                     binwidth = 0.02,
+                     scale = TRUE,
+                     neighbour_limit = 1,
+                     method = "density",
+                     adjust = 1,
+                     na.rm = FALSE,
+                     show.legend = NA,
+                     inherit.aes = TRUE) {
+  method <- match.arg(method, c("density", "neighbourhood"))
+
+  layer(
+    data = data,
+    mapping = mapping,
+    stat = StatSina,
+    geom = geom,
+    position = position,
+    show.legend = show.legend,
+    inherit.aes = inherit.aes,
+    params = list(
+      binwidth = binwidth,
+      scale = scale,
+      neighbour_limit = neighbour_limit,
+      method = method,
+      adjust = adjust,
+      na.rm = na.rm,
+      ...
+      )
+  )
+}
+
+
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
+#' @export
+StatSina <- ggproto("StatSina", Stat,
+
+  required_aes = c("x", "y"),
+
+  setup_data = function(data, params) {
+
+    if (is.double(data$x) && !has_groups(data) && any(data$x != data$x[1L])) {
+      stop(
+        "Continuous x aesthetic -- did you forget aes(group=...)?",
+        call. = FALSE)
+    }
+
+    data
+  },
+
+  setup_params = function(data, params) {
+
+    #scale adjust value
+    if (params$method == "neighbourhood"){
+      if (!is.null(params$adjust))
+        params$adjust <- params$adjust / 100
+      else
+        params$adjust <- 0.01
+    }
+
+    params
+  },
+
+
+  compute_panel = function(self, data, scales, binwidth = 0.02, scale = TRUE,
+                           neighbour_limit = 1, method = "d", adjust = 1,
+                           na.rm = FALSE) {
+
+    bins <- bin_y(data$y, binwidth)
+
+    data <- ggproto_parent(Stat, self)$compute_panel(data, scales,
+      binwidth = binwidth, scale = scale, neighbour_limit = neighbour_limit,
+      method = method, adjust = adjust, bins = bins, na.rm = na.rm)
+
+    #scale all neighbourhoods based on their density relative to the
+    #densiest neighbourhood
+    if (scale)
+      group_scaling_factor <- data$max_neighbours / max(data$max_neighbours)
+    else
+      group_scaling_factor <- 1
+
+    #translate x coordinates
+    data$x <- data$x + data$x_translation * group_scaling_factor
```

Comment:
It is better to create a new column

```R
data$scaled <- data$x + data$x_translation * group_scaling_factor

# maybe even
data$unscaled <- data$x + data$x_translation

# then delete
data$x_translation <- NULL
```
Every column left in there becomes a computed aesthetic.